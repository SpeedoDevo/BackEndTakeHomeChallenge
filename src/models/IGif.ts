interface Data {
  title: string
  url: string
}

export default interface IGif {
  toObject(): Data
}
