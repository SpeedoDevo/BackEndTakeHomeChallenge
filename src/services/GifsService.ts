import request from 'request-promise-native'
import Gif from '../models/Gif'
import IGif from '../models/IGif'
import IGiphy from '../models/IGiphy'

const BASE_URL = 'https://api.giphy.com/v1'
// this really shouldn't be in VCS
const API_KEY = 'YQD46PNuiExynupQ46HOzJqctivOHt16'

export default class GifsService {
  async getTrending(): Promise<Array<IGif>> {
    const result: { data: IGiphy[] } = await request(
      `${BASE_URL}/gifs/trending`,
      {
        qs: {
          api_key: API_KEY,
        },
        json: true,
      },
    )

    return result.data.map(Gif.fromGiphy)
  }
}
