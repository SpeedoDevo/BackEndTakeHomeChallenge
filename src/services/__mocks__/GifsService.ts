import IGif from '../../models/IGif'

const mockData = [
  {
    title: 'Sea Otter Good Job GIF by Justin',
    url:
      'https://media1.giphy.com/media/ely3apij36BJhoZ234/giphy-downsized-large.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy-downsized-large.gif',
  },
  {
    title: 'Yes GIF by Brittany Broski',
    url:
      'https://media2.giphy.com/media/lnOG1o6Cdc3kKjuray/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'Spongebob Squarepant GIF by The SpongeBob Movie: Sponge On The Run',
    url:
      'https://media1.giphy.com/media/PnUatAYWMEMvmiwsyx/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'happy dog GIF by TikTok',
    url:
      'https://media3.giphy.com/media/1BGQQpzJ0N8K99RgzB/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'Ashley Graham Lol GIF by Lifetime',
    url:
      'https://media0.giphy.com/media/3qsz090iknpyrnIXrX/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'everybody wants to be a cat happy dance GIF',
    url:
      'https://media2.giphy.com/media/EkPicGgIeyNAQ/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'Season 4 GIF by Rick and Morty',
    url:
      'https://media2.giphy.com/media/PlTd840Mkz4BaiUw05/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'trippy processing GIF',
    url:
      'https://media2.giphy.com/media/ExYQhwnfUzgLm/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'freaks and geeks dancing GIF by Justin',
    url:
      'https://media0.giphy.com/media/l4Ep6Ee5dBpOB44SI/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'funny baby GIF',
    url:
      'https://media1.giphy.com/media/YAXXp5UguFaQQIJzPH/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'the fresh prince of bel air dancing GIF',
    url:
      'https://media3.giphy.com/media/pa37AAGzKXoek/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'doubt GIF by Samm Henshaw',
    url:
      'https://media0.giphy.com/media/1d5WSz56LVnDtyb0Go/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'Minnesota Timberwolves Reaction GIF by Milwaukee Bucks',
    url:
      'https://media2.giphy.com/media/gjHKcCYscYE14o7rUU/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'Come On Reaction GIF by WWE',
    url:
      'https://media0.giphy.com/media/jUKvCSdBaaLD9saKHT/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'Star Wars Baby Yoda GIF by melbduran',
    url:
      'https://media0.giphy.com/media/ZFi2wFRs3lBvYNglWk/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'Happy Clapping GIF by Originals',
    url:
      'https://media3.giphy.com/media/l0K4m0mzkJDAIdhHW/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'Trail Blazers No GIF by Twitter',
    url:
      'https://media2.giphy.com/media/iddDVZCrCDRva83r1b/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: '',
    url:
      'https://media1.giphy.com/media/iIe5ficubsfcYj0mxo/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'give valentines day GIF by Originals',
    url:
      'https://media3.giphy.com/media/XHxW40d07qdxsf6U52/giphy-downsized-large.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy-downsized-large.gif',
  },
  {
    title: 'feel better get well soon GIF',
    url:
      'https://media2.giphy.com/media/SHyuhBtRr8Zeo/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'steal yo girl GIF by Portland Trail Blazers',
    url:
      'https://media3.giphy.com/media/26uf4CdbJ2uPxFjMI/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'GIF by NBA',
    url:
      'https://media2.giphy.com/media/TKGOR4RxYlFgqZRui1/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'Cat Shoulder Touch GIF by Nebraska Humane Society',
    url:
      'https://media0.giphy.com/media/S857VNxM6HwlZuYXrU/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'GIF by NBA',
    url:
      'https://media2.giphy.com/media/QBwPwGhGIqWUGBNEUq/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
  {
    title: 'Shocked Laura Dern GIF by Big Little Lies',
    url:
      'https://media3.giphy.com/media/UokMtxPBdBn2Da3Xzs/giphy.gif?cid=f1afa41d1726315b8ee94c63848d339c6679d5e17adafa12&rid=giphy.gif',
  },
]

export default class MockGifsService {
  async getTrending(): Promise<IGif[]> {
    return mockData.map(o => ({ toObject: () => o }))
  }
}
