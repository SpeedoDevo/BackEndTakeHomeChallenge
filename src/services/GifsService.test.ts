import _request from 'request-promise-native'
import IGif from '../models/IGif'
import GifsService from './GifsService'

const request = (_request as unknown) as jest.Mock<Promise<any>, any[]>
jest.mock('request-promise-native')

describe('GifsService', () => {
  let service: GifsService

  beforeEach(() => {
    service = new GifsService()
  })

  describe('getTrending', () => {
    let res: IGif[]

    beforeEach(async () => {
      request.mockResolvedValue({
        data: [
          {
            title: 'a gif 1',
            images: { downsized_large: { url: '/the-url-1' } },
          },
          {
            title: 'a gif 2',
            images: { downsized_large: { url: '/the-url-2' } },
          },
          {
            title: 'a gif 3',
            images: { downsized_large: { url: '/the-url-3' } },
          },
        ],
      })
      res = await service.getTrending()
    })

    it('should make a request to giphy', () => {
      expect(request).toHaveBeenCalledWith(
        'https://api.giphy.com/v1/gifs/trending',
        {
          json: true,
          qs: { api_key: 'YQD46PNuiExynupQ46HOzJqctivOHt16' },
        },
      )
    })

    it('should create new model objects', () => {
      expect(res.map(gif => gif.toObject())).toMatchSnapshot()
    })
  })
})
