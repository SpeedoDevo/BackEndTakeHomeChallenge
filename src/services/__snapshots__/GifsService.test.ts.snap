// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifsService getTrending should create new model objects 1`] = `
Array [
  Object {
    "title": "a gif 1",
    "url": "/the-url-1",
  },
  Object {
    "title": "a gif 2",
    "url": "/the-url-2",
  },
  Object {
    "title": "a gif 3",
    "url": "/the-url-3",
  },
]
`;
